src/test/scala/is/hail/methods/SkatSuite.scala

```diff
@@ -0,0 +1,484 @@
+    var i = 0
+
+    while (i < resultsArray.size) {
+
+      val qstat = rows(i).get(1).asInstanceOf[Double]
+      val pval = rows(i).get(2).asInstanceOf[Double]
+
+      val qstatR = resultsArray(i)(1).asInstanceOf[Double]
+      val pvalR = resultsArray(i)(2).asInstanceOf[Double]
+      if (pval <= 1 && pval >= 0) {
+
+        assert(D_==(qstat, qstatR, tol))
+        assert(D_==(pval, pvalR, tol))
+      }
+
+      i += 1
+    }
+
+  }
+
+  @Test def dosagesSmallTest() {
+
+    val useDosages = true
+
+    val resultsArray = testInR(vds, "gene", "va.genes", singleKey = false,
+      Some("va.weight"), "sa.pheno", Array("sa.cov.Cov1", "sa.cov.Cov2"), useDosages)
+
+    val kt = vds.skat("gene", "va.genes", singleKey = false, Option("va.weight"),
+      "sa.pheno", covariates = Array("sa.cov.Cov1", "sa.cov.Cov2"), useDosages)
+
+
+    val rows = kt.rdd.collect()
+
+    val tol = 1e-5
+    var i = 0
+
+    while (i < resultsArray.size) {
+      val qstat = rows(i).get(1).asInstanceOf[Double]
+      val pval = rows(i).get(2).asInstanceOf[Double]
+
+      val qstatR = resultsArray(i)(1).asInstanceOf[Double]
+      val pvalR = resultsArray(i)(2).asInstanceOf[Double]
+
+      if (pval <= 1 && pval >= 0) {
+        assert(D_==(qstat, qstatR, tol))
+        assert(D_==(pval, pvalR, tol))
+      }
+      i += 1
+    }
+
+  }
+
+
+  @Test def largeNSmallTest() {
+
+    val useDosages = false
+    val useLargeN = true
+
+    val kt = Skat(vds, "gene", "va.genes", false, Option("va.weight"), "sa.pheno",
+      Array("sa.cov.Cov1", "sa.cov.Cov2"),useDosages, useLargeN)
+
+    val resultsArray = testInR(vds, "gene", "va.genes", singleKey = false,
+      Some("va.weight"), "sa.pheno", Array("sa.cov.Cov1", "sa.cov.Cov2"), useDosages)
+
+    val rows = kt.rdd.collect()
+
+    val tol = 1e-5
+    var i = 0
+
+    while (i < resultsArray.size) {
+      val qstat = rows(i).get(1).asInstanceOf[Double]
+      val pval = rows(i).get(2).asInstanceOf[Double]
+
+      val qstatR = resultsArray(i)(1).asInstanceOf[Double]
+      val pvalR = resultsArray(i)(2).asInstanceOf[Double]
+
+
+      if (pval <= 1 && pval >= 0) {
+        assert(D_==(qstat, qstatR, tol))
+        assert(D_==(pval, pvalR, tol))
+      }
+      i += 1
+    }
+
+  }
+
+
+  //Generates random data
+  def buildWeightValues(filepath: String): Unit = {
+
+    //read in chrom:pos values
+    val fileSource = Source.fromFile(filepath)
+    val linesArray = fileSource.getLines.toArray
+    fileSource.close()
+
+    //write in randomized weights
+    val fileObject = new PrintWriter(new File(filepath))
+
+    for (i <- -1 until linesArray.size) {
+      if (i == -1) {
+        fileObject.write("Pos\tWeights\n")
+      }
+      else {
+        val pos = linesArray(i)
+        val randomWeight = scala.util.Random.nextDouble()
+        fileObject.write(s"$pos\t$randomWeight\n")
+
+      }
+    }
+    fileObject.close()
+  }
+
+  def buildCovariateMatrix(filepath: String, covariateCount: Int): Unit = {
```

Comment:
It's not used in the testing suite when it's run, it was just written for building test cases by generating random covariate matrices.